packages/graphql/src/services/sessions.ts

```diff
@@ -383,20 +446,93 @@ export async function deactivateSessionBlock(
       blocks: {
         update: {
           where: {
-            id: sessionBlockId,
+            id: Number(sessionBlockId),
           },
           data: {
             status: SessionBlockStatus.EXECUTED,
             instances: {
-              update: [],
+              update: Object.entries(instanceResults).map(([id, results]) => ({
+                where: { id: Number(id) },
+                data: {
+                  results: results.results,
+                  // TODO: persist responses or "too much information"? delete when session is completed? what about anonymous users?
+                  responses: {
+                    create: Object.entries(results.responses).map(
+                      ([participantId, response]) => ({
+                        response,
+                        participant: {
+                          connect: { id: participantId },
+                        },
+                        participation: {
+                          connect: {
+                            courseId_participantId: {
+                              // TODO: this is not set if the session is not in a course (i.e., not gamified)
+                              courseId: session.courseId as string,
+                              participantId,
+                            },
+                          },
+                        },
+                      })
+                    ),
+                  },
+                },
+              })),
+            },
+            leaderboard: {
+              create: Object.entries(blockLeaderboard).map(([id, score]) => ({
+                score: Number(score),
+                participant: {
+                  connect: { id },
+                },
+                type: 'SESSION_BLOCK',
+                // TODO: use the real username here, or remove it
+                username: id,
+              })),
             },
           },
         },
       },
+      leaderboard: {
+        upsert: Object.entries(sessionLeaderboard).map(([id, score]) => ({
+          where: {
+            type_participantId_sessionId: {
+              type: 'SESSION',
+              participantId: id,
+              sessionId,
+            },
+          },
+          create: {
+            type: 'SESSION',
+            participant: {
+              connect: { id },
+            },
+            score: Number(score),
+            // TODO: use the real username here, or remove it
+            username: id,
+          },
+          update: {
+            score: Number(score),
+          },
+        })),
+      },
+    },
+    include: {
+      blocks: true,
     },
   })
 
-  return session
+  // unlink everything regarding the block in redis
+  const unlinkMulti = ctx.redisExec.pipeline()
```

Comment:
unlink deletes the irrelevant stuff in the redis cache (as it has been persisted to the db above). but it does so asynchronously, in a separate thread on redis (i.e., if the cache is not under load, it deletes it)